utility/AlwaysAssert.hpp

```diff
@@ -0,0 +1,54 @@
+/*
+* Copyright (c) 2015, Intel Corporation
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without modification,
+* are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+* list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+* this list of conditions and the following disclaimer in the documentation and/or
+* other materials provided with the distribution.
+*
+* 3. Neither the name of the copyright holder nor the names of its contributors
+* may be used to endorse or promote products derived from this software without
+* specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+* ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+* ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include <iostream>
+
+
+#ifdef NDEBUG
+#    include <exception>
+#    define ALWAYS_ASSERT_FAILURE(cond) std::terminate()
+#else
+#    include <cassert>
+#    define ALWAYS_ASSERT_FAILURE(cond) assert(cond)
+#endif
+
+// __func__ should be used as per C99 and C++11, but as visual studio 2013 isn't
+// compliant, use __FUNCTION__ which is compatible with linux and WIN32.
+
+
+#define ALWAYS_ASSERT(cond, iostr) \
+    do { \
+         if (!(cond)) { \
+           std::cerr << __FILE__ << ":" << __LINE__ << ":" << __FUNCTION__ \
+           << ": Assert '" #cond "' failed: " << iostr << std::endl; \
+            ALWAYS_ASSERT_FAILURE(cond); \
+         } \
```

Comment:
Coding style would recommend:
``` cpp
do { \
    if (!(cond)) { \
        std::cerr << __FILE__ << ":" << __LINE__ << ":" << __FUNCTION__ \
                  << ": Assert '" #cond "' failed: " << iostr << std::endl; \
        ALWAYS_ASSERT_FAILURE(cond); \
    } \
} \
```
 - << should be allign, 
 - do -> if only 4 space
but if you do not want, I would understand. :+1: